Use faster sort in TopK kernel when num_rows==1

benbarsdell opened the pull request:
`DeviceSegmentedRadixSort` is very slow when `num_segments=1` because it only uses 1 SM per segment. Calling the un-segmented version is much faster (3-5x) in this case (i.e., for 1D input).

No functional change besides the improved performance.

cc @nluehr

benbarsdell committed:
Use faster sort in TopK kernel when num_rows==1

- DeviceSegmentedRadixSort is very slow when num_segments=1 because it
  only uses 1 SM per segment. Calling the un-segmented version is much
  faster (3-5x) in this case.

gbaned commented:
@jpienaar Can you please review this PR ? Thank you!

benbarsdell commented:
Gentle ping.

jpienaar commented:
Ah sorry, just saw this (was OOO). I'm not sure why assigned to me, let me see if I can find owner here. Sorry for delay this slipped through.

jpienaar commented:
Mmm, looks like the person I'd want to rope in is not accessible. I'll ask XLA GPU folks as it's been years since I've done GPU programming (either to review or redirect).

cheshire commented:
@reedwm Could you take a look?

gbaned commented:
@benbarsdell Can you please address Ubuntu Sanity errors? Thank you!

benbarsdell commented:
Fixed and rebased.

Note that the ROCm failure looks like it's related to this:
https://github.com/ROCmSoftwarePlatform/rocPRIM/commit/e6ee421eec08f046841b3924c613a3efd919be74
